### PR TITLE
feat(pod): change pod boolean props UI to switch button (issue #243)

### PR DESCRIPTION
**Pull request by franquil**
My first PR which targets to fix the issue #243 .

A couple of things need to be clarified:

- Two UI frameworks "react-bootstrap" and "shards-react" are being used in the repo, which is a bit overlapping; I wonder which one is recommended?
- Are there any tests that I could add/update in terms of the feature targeted in the PR? I couldn't see the tests regarding this specific component. Where can I find them? Does this PR require to add integration tests?

Thanks in advance.

<h2>PR checklist</h2>

**Docs**
- [x] If you added new code; is it self documenting? (all the variable names clear, comments added where appropriate)
- [ ] If you touched existing code; did you refactor existing naming, added comments where necessary?

**Tests**
- [ ] If your ticket adds new functionality; have you added unit tests to verify behaviour?
- [ ] If your ticket alters user interaction with the UI; have you added integration tests?
- [ ] Do your tests check for every _convievable_ user behabviour? (all behaviours the user _can_ perform)

**Typing**
- [ ] If you added new code; have you added typing?
- [ ] If you touched existing code; have you improved its typing?

**Styling**
- [ ] If new styles were created; Did you refactor/add more styles in css-in-js style?

### CODE REVIEW

**Comment by franquil** [comment]
Does anyone have an idea about why the checks failed at 'ref-issue-to-pr'?
Error:
```
Error: GraphqlError: Resource not accessible by integration
Error: Resource not accessible by integration
```

**Comment by franquil** [comment]
> Does anyone have an idea about why the checks failed at 'ref-issue-to-pr'?
> Error:
> 
> ```
> Error: GraphqlError: Resource not accessible by integration
> Error: Resource not accessible by integration
> ```

The error was gone after a new submitting.

**Comment by BastinJafari** [comment]
Hey @franquil, thank you very much for your interest in contributing!

Regarding UI Frameworks, we want to switch to Material, so it would be nice to use that.
Regarding Testing. We use Cypress for integration tests. If you like, we can schedule a Pair Programming session and I can show you how to do it. You can write me in Jina's Community Slack under Bastin Jafari

**Comment by franquil** [comment]
> Hey @franquil, thank you very much for your interest in contributing!
> 
> Regarding UI Frameworks, we want to switch to Material, so it would be nice to use that.
> Regarding Testing. We use Cypress for integration tests. If you like, we can schedule a Pair Programming session and I can show you how to do it. You can write me in Jina's Community Slack under Bastin Jafari

Hi @BastinJafari , thanks for your reply. I notice Bootstrap rather than Material framework is used in this repo. It sounds like a much bigger task, is there a road map or open issues concerning it? 

In terms of the testing, is there a plan to adopt a component-level (or unit) test? I tried Cypress tests locally, there's a timeout issue when connecting local server that consistently occurred.  I'd like to have a look at it on another day. Thanks for offering your help.

By the way, could you or someone else review this PR?

Thanks.

**Comment by BastinJafari** [comment]
We have a roadmap:

http://roadmap.jina.ai/

it doesn't mention the use of MaterialUI though.

I would say forget the integration of Material, for now. We should do that.
But this PR still needs integration tests. Can you make a screenshot of the error or paste it?

**Comment by franquil** [comment]
> We have a roadmap:
> 
> http://roadmap.jina.ai/
> 
> it doesn't mention the use of MaterialUI though.
> 
> I would say forget the integration of Material, for now. We should do that.
> But this PR still needs integration tests. Can you make a screenshot of the error or paste it?

@BastinJafari  Here's the process to reproduce the errors:
```
docker run -p 8000:8000 -p 5555:5555 jinaai/jina:master-daemon
cd dashboard
npm run dev
npm run cy:run
```
My laptop environment:
```
# docker
Docker version 19.03.6
# linux
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal
```

The error:
```
> dashboard-v1@1.1.0 cy:run
> percy exec -- npx cypress run

 ›   Warning: Skipping visual tests. PERCY_TOKEN was not provided.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    6.8.0                                                                              │
  │ Browser:    Electron 87 (headless)                                                             │
  │ Specs:      6 found (views/flowview.ts, views/helpView.ts, views/hubView.ts, views/logsView.ts │
  │             , visual-tests/flowDesign.ts, visual-tests/helpPage.ts)                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  views/flowview.ts                                                               (1 of 6)


  The Flow Page
    1) "before each" hook for "should have a working settings button"
    2) "after each" hook for "should have a working settings button"


  0 passing (1m)
  2 failing

  1) The Flow Page
       "before each" hook for "should have a working settings button":
     CypressError: Timed out after waiting `60000ms` for your remote page to load.

Your page did not fire its `load` event within `60000ms`.

You can try increasing the `pageLoadTimeout` value in `/home/frank/playground/jina/dashboard/cypress.json` to wait longer.

Browsers will not fire the `load` event until all stylesheets and scripts are done downloading.

When this `load` event occurs, Cypress will continue running commands.

Because this error occurred during a `before each` hook we are skipping the remaining tests in the current suite: `The Flow Page`
      at timedOutWaitingForPageLoad (http://localhost:3000/__cypress/runner/cypress_runner.js:158187:13)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:159171:16
      at tryCatcher (http://localhost:3000/__cypress/runner/cypress_runner.js:10765:23)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:5887:41
      at tryCatcher (http://localhost:3000/__cypress/runner/cypress_runner.js:10765:23)
      at Promise._settlePromiseFromHandler (http://localhost:3000/__cypress/runner/cypress_runner.js:8700:31)
      at Promise._settlePromise (http://localhost:3000/__cypress/runner/cypress_runner.js:8757:18)
      at Promise._settlePromise0 (http://localhost:3000/__cypress/runner/cypress_runner.js:8802:10)
      at Promise._settlePromises (http://localhost:3000/__cypress/runner/cypress_runner.js:8878:18)
      at _drainQueueStep (http://localhost:3000/__cypress/runner/cypress_runner.js:5472:12)
      at _drainQueue (http://localhost:3000/__cypress/runner/cypress_runner.js:5465:9)
      at Async.../../node_modules/bluebird/js/release/async.js.Async._drainQueues (http://localhost:3000/__cypress/runner/cypress_runner.js:5481:5)
      at Async.drainQueues (http://localhost:3000/__cypress/runner/cypress_runner.js:5351:14)
  From Your Spec Code:
      at Context.eval (http://localhost:3000/__cypress/tests?p=cypress/integration/views/flowview.ts:185:12)

  2) The Flow Page
       "after each" hook for "should have a working settings button":
     CypressError: Cypress command timeout of `4000ms` exceeded.

Because this error occurred during a `after each` hook we are skipping all of the remaining tests.
      at cypressErr (http://localhost:3000/__cypress/runner/cypress_runner.js:174213:18)
      at Object.errByPath (http://localhost:3000/__cypress/runner/cypress_runner.js:174264:10)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:175857:29




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        5                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      4                                                                                │
  │ Screenshots:  2                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     1 minute, 4 seconds                                                              │
  │ Spec Ran:     views/flowview.ts                                                                │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Similar errors occurred at the Help Page, Hub Page and Log Page etc. Taking the error of Help Page as an example:
```
 The Help Page
    1) "before all" hook for "should redirect to /help"

  2) "after all" hook: collectBackendCoverage for "should redirect to /help"

  0 passing (1m)
  2 failing

  1) The Help Page
       "before all" hook for "should redirect to /help":
     CypressError: Timed out after waiting `60000ms` for your remote page to load.

Your page did not fire its `load` event within `60000ms`.

You can try increasing the `pageLoadTimeout` value in `/home/frank/playground/jina/dashboard/cypress.json` to wait longer.

Browsers will not fire the `load` event until all stylesheets and scripts are done downloading.

When this `load` event occurs, Cypress will continue running commands.

Because this error occurred during a `before all` hook we are skipping the remaining tests in the current suite: `The Help Page`
      at timedOutWaitingForPageLoad (http://localhost:3000/__cypress/runner/cypress_runner.js:158187:13)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:159171:16
      at tryCatcher (http://localhost:3000/__cypress/runner/cypress_runner.js:10765:23)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:5887:41
      at tryCatcher (http://localhost:3000/__cypress/runner/cypress_runner.js:10765:23)
      at Promise._settlePromiseFromHandler (http://localhost:3000/__cypress/runner/cypress_runner.js:8700:31)
      at Promise._settlePromise (http://localhost:3000/__cypress/runner/cypress_runner.js:8757:18)
      at Promise._settlePromise0 (http://localhost:3000/__cypress/runner/cypress_runner.js:8802:10)
      at Promise._settlePromises (http://localhost:3000/__cypress/runner/cypress_runner.js:8878:18)
      at _drainQueueStep (http://localhost:3000/__cypress/runner/cypress_runner.js:5472:12)
      at _drainQueue (http://localhost:3000/__cypress/runner/cypress_runner.js:5465:9)
      at Async.../../node_modules/bluebird/js/release/async.js.Async._drainQueues (http://localhost:3000/__cypress/runner/cypress_runner.js:5481:5)
      at Async.drainQueues (http://localhost:3000/__cypress/runner/cypress_runner.js:5351:14)
  From Your Spec Code:
      at Context.eval (http://localhost:3000/__cypress/tests?p=cypress/integration/views/helpView.ts:98:12)

  2) The Help Page
       "after all" hook: collectBackendCoverage for "should redirect to /help":
     CypressError: Cypress command timeout of `4000ms` exceeded.

Because this error occurred during a `after all` hook we are skipping all of the remaining tests.
      at cypressErr (http://localhost:3000/__cypress/runner/cypress_runner.js:174213:18)
      at Object.errByPath (http://localhost:3000/__cypress/runner/cypress_runner.js:174264:10)
      at http://localhost:3000/__cypress/runner/cypress_runner.js:175857:29




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        14                                                                               │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      13                                                                               │
  │ Screenshots:  2                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     1 minute, 4 seconds                                                              │
  │ Spec Ran:     views/helpView.ts                                                                │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
````

Occationally, some of the tests can be passed. But most of times the timeout issue would occur.

Thanks for offering your help.

**Comment by BastinJafari** [comment]
It's weird, I can't reproduce the error on your branch on my machine. 
The page under localhost:3000 normally loads?

Could you try `npm run cy:dev` instead and see if you can learn anything from that?

**Comment by franquil** [comment]
> It's weird, I can't reproduce the error on your branch on my machine.
> The page under localhost:3000 normally loads?
> 
> Could you try `npm run cy:dev` instead and see if you can learn anything from that?

Hi @BastinJafari , I figured out what caused the timeout issue, now the tests can run as normal. It seems that there're external resources (hosted on cloudfalre or googleapi etc) that failed to load in time, after setting `HTTP_PROXY`, the errors are gone. (By the way, I'm based in China. The connection to some worldwide servers are not stable sometimes.)

A couple of integration tests are added. Let me know if there're changes needed. Thanks.

**Comment by Roshanjossey** [comment]
Hi @franquil, good job figuring out the issue. Yeah, we have resources like fonts, icons etc hosted on CDNs. But now, I see value in packaging it along with the build.

File: cypress/integration/views/podView.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  it("should set default pod's property input as checkbox whichever its type is boolean, and make it checkable", () => {
+    const booleanProperties = flowArguments.pod.filter(
+      (arg) => arg.type === "boolean"
+    )
+    cy.dataName("chart-node-gateway").dblclick()
+    booleanProperties.forEach((prop) => {
+      cy.get(`input[data-name=pod_${prop.name}_property][type=checkbox]`)
+        .as(prop.name)
+        .should("not.be.checked")
+      cy.get(`@${prop.name}`).check({ force: true }).should("be.checked")
```
Comment (Roshanjossey) [inline cypress/integration/views/podView.ts:30]: @franquil, did you get an error around element being detached from DOM when you use the checkbox without `force: true` ?

@BastinJafari, in my observation, this detaching problem only happens in Flow page. We should spend some time to get to the bottom of it.

Comment (franquil) [reply cypress/integration/views/podView.ts:30]: @Roshanjossey thanks for your instant reply. I did encounter errors if `{force: true}` is removed. I suppose in this case it's caused by the target element which is not visible (with style `position: fixed`). I'm not sure it's the same for the Flow page. There're some cases according to the [official guide](https://docs.cypress.io/guides/core-concepts/interacting-with-elements#Forcing), the `force` attribute can be very useful.

Comment (Roshanjossey) [reply cypress/integration/views/podView.ts:30]: @franquil, I don't think `position: fixed` is the problem here. It's more likely that React updated the DOM and the element that was selected before is not in the current DOM. I could be wrong here.

Anyways, I think it's good to investigate this and understand what's going on.

File: src/helpers/tests/flow-chart.testData.ts [hunk 2]
```diff
@@ -223,7 +223,7 @@ export const flowArguments: FlowArguments = {
       type: "string",
     },
     {
-      name: "hide_exc_info",
```
Comment (BastinJafari) [inline src/helpers/tests/flow-chart.testData.ts:226]: why did you change the name here?

File: src/modals/PodEdit.tsx [hunk 3]
```diff
@@ -141,7 +142,7 @@ function PodEditComponent({ open, closeModal, modalParams }: Props) {
       closeTimeoutMS={100}
       style={style}
     >
-      <PodEditContainer>
+      <PodEditContainer data-name="pod-edit-container">
```
Comment (BastinJafari) [inline src/modals/PodEdit.tsx:145]: We use camelCase for data-name

**Review by BastinJafari — CHANGES_REQUESTED** [review]

**Comment by BastinJafari** [comment]
Very nice PR!
Just had a question and some minor comments

File: src/helpers/tests/flow-chart.testData.ts (hunk 2 above)
Comment (franquil) [reply src/helpers/tests/flow-chart.testData.ts:226]: Two reasons: 1) according to the meaning of the option's description "If set, then exception stack information to be added to the logging message, useful in debugging", which implies the name should be `show_exc_info` instead of `hide_exc_info`; 2) based on the UI of Pod Edit Modal, it shows the option as "show_exc_info". 

If the option suppose to be `hide_exc_info`, the description and the text on Pod Edit Modal should be corrected. I wonder which one the option does indicate, and whether the default behaviour is showing or hiding the exc info. 

Thanks.

File: src/modals/PodEdit.tsx (hunk 3 above)
Comment (franquil) [reply src/modals/PodEdit.tsx]: Thanks for letting me know. The attribute is updated.

File: src/helpers/tests/flow-chart.testData.ts (hunk 2 above)
Comment (imsergiy) [reply src/helpers/tests/flow-chart.testData.ts:226]: One quick thing, this is test data so these exact flow arguments won't be the ones displaying in the dashboard. We fetch flow arguments from jinaAPI or jinad based on connection status, we don't have control over these arguments or their naming in the dashboard. We should leave this unchanged as it is a copy of what we pulled from jina api.

Comment (franquil) [reply src/helpers/tests/flow-chart.testData.ts:226]: Hi @imsergiy , thanks for your comment. Currently, I suppose there's a mocked server which located at `dev-resources` directory, and all the flow arguments are pulled from there?  The mocked argument name for the input is `show_exc_info`, I guess that's why at the dashboard Pod Edit Modal it shows `show_exc_info`.

The argement's name at mocked server is different from the one at the `flow-chart.testData.ts` file, which caused a bit trouble, because at the cypress `podView.ts` test, the argument's name is used for targeting the input element. If the name is set as `hide_exc_info`, Cypress won't find it, due to the current argument's name from mocked server is 'show_exc_info'.

If the mocked server's response is the same with the ones at `flow-chart.testData.ts`, would it be a bit easier to understand and test? To keep the argument identical, I can either change the argument name at `flow-chart.testData.ts`  and mocked server to `hide_exc_info`, or change them to `show_exc_info`?

What do you guys think? Thanks. @BastinJafari @imsergiy

Comment (imsergiy) [reply src/helpers/tests/flow-chart.testData.ts:226]: I see, good point. Either way works then since it's just mock data.

Comment (imsergiy) [reply src/helpers/tests/flow-chart.testData.ts:226]: It would be good to keep it consistent between testServer and dashboard, I like your idea.

Comment (franquil) [reply src/helpers/tests/flow-chart.testData.ts]: Hi @imsergiy @BastinJafari , to keep the data consistent, the argument's name `show_exc_info` is changed to `hide_exc_info`. However, according to the argument's description "If set, then exception stack information to be added to the logging message, useful in debugging", I'm a bit confused about the expected default behaviour, whether the info will be logged or not. Anyway, I suppose it can be fixed at another PR.

**Review by Roshanjossey — APPROVED** [review]

**Review by imsergiy — APPROVED** [review]
